python/ray/util/sgd/torch/ptl_operator.py

```diff
@@ -0,0 +1,479 @@
+import inspect
+import logging
+
+import torch
+from pytorch_lightning.core.step_result import Result
+from pytorch_lightning.overrides.data_parallel import \
+    LightningDistributedDataParallel
+from pytorch_lightning.trainer.model_hooks import TrainerModelHooksMixin
+from pytorch_lightning.trainer.optimizers import TrainerOptimizersMixin
+import pytorch_lightning as ptl
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.memory import recursive_detach
+from ray.util.sgd.torch import TrainingOperator
+from ray.util.sgd.torch.constants import NUM_STEPS, SCHEDULER_STEP_BATCH, \
+    SCHEDULER_STEP_EPOCH
+from ray.util.sgd.utils import AverageMeterCollection, NUM_SAMPLES
+
+tqdm = None
+try:
+    from tqdm import tqdm
+except ImportError:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+class PTLOperator(TrainingOperator, TrainerModelHooksMixin,
+                  TrainerOptimizersMixin):
+    def _configure_amp(self, amp, models, optimizers):
+        assert len(models) == 1
+        model = models[0]
+        assert isinstance(model, ptl.LightningModule)
+        amp_level = self._apex_args.get("opt_level", "O2")
+        model, optimizers = model.configure_apex(
+            amp, model, optimizers, amp_level=amp_level)
+        return [model], optimizers
+
+    def _configure_ddp(self, models, device_ids):
+        assert len(models) == 1
+        model = models[0]
+        assert isinstance(model, ptl.LightningModule)
+        # This will default to LightningDistributedDataParallel.
+        model = model.configure_ddp(model=model, device_ids=device_ids)
+        return [model]
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def scheduler_dicts(self):
+        return self._scheduler_dicts
+
+    @property
+    def optimizers(self):
+        return self._optimizers
+
+    @property
+    def schedulers(self):
+        return self._schedulers
+
+    def get_model(self):
+        if isinstance(self.model, LightningDistributedDataParallel):
+            return self.model.module
+        else:
+            return self.model
+
+    def setup(self, config):
+        self._ptl = True
+        # Pass in config if ptl_module accepts it.
+        ptl_class = self.__class__._lightning_module_cls
+        if "config" in inspect.signature(ptl_class.__init__).parameters:
+            ptl_module = ptl_class(config=config)
+        else:
+            ptl_module = ptl_class()
+        if not isinstance(ptl_module, ptl.LightningModule):
```

Comment:
you can do a `issubclass` check on the `_lightning_module_cls` provided - that'll help you raise TypeError earlier

Reply:
ah this is great! made the change.